src/branch.c

```diff
@@ -597,97 +597,99 @@ static int unset_upstream(git_config *config, const char *shortname)
 	return -1;
 }
 
-int git_branch_set_upstream(git_reference *branch, const char *upstream_name)
+int git_branch_set_upstream(git_reference *branch, const char *branch_name)
```

Comment:
To not loose the upstream-y nature, what about `upstream_branch`? The other renames are definitely helpful, though.

Reply:
Not sure. My reasoning was that "upstream" is only meaningful when it's actually set, and that you call this function with a branch handy. To me, "upstream" muddies the waters by not being really what's needed here, ie, you could take a `git_remote *` object, and it'd make more sense.

Maybe that's what I'm misunderstanding — so, just to make sure I'm not, here's my "glossary":
- a "branch" is a reference in some "well-known" refdb location (either `refs/heads/` for a local/`"."` branch, or `refs/remotes/`).
- its "branch name" is the part under `refs/heads/` for a local branch, or `refs/remotes/*/` for a remote one.
- a "tracking branch" has a config section, that specifies its "upstream configuration", lives in `refs/heads/` (not sure if that's always true), and should (ideally) resolve to another reference under those configuration rules.
- "upstream" is the "logical" remote, known via the "rev-refspec" resolution.
- "upstream name" is the rev-refspec branch name (the ref's shortname, a.k.a. the refname, offset by "refs/remote/$rev-refspec/").
- "upstream branch" is the reference with that full rev-refspec name, a.k.a. the branch's tracking reference/branch.

I think what's confusing is that a branch is both "a reference name" at lookup, but "a name from a config key" later on, and the API reflects this confusion. IMHO, this could be fixed by just making `git_branch_lookup` more consistent — say `git_branch_lookup(git_reference **out, git_repository *repo, const char *branch_name, git_remote *remote)`. Hence, you pass a `NULL` remote for a "local, maybe tracking" branch, and remote ones can be checked for validity using the refspecs.

Reply:
Except for upstream name I think you've got everything right. For "upstream name" itself, I don't think that it actually has a well defined meaning.

The thing that bothers me is that `branch` and `branch_name` are so similar to each other, giving some false signals at first. The thing is we _are_ about to set the upstream in this function, right? That's why I thought that `upstream_branch` would be fine. I'm definitely open to better proposals, though.

Reply:
> Except for upstream name I think you've got everything right. For "upstream name" itself, I don't think that it actually has a well defined meaning.

I'm clearly unsure as well, but it seems to describe what's provided, so 👍 ? "Upstream branch name" would be a more clear version, but it's a mouthful…

> The thing that bothers me is that branch and branch_name are so similar to each other, giving some false signals at first.

Agreed. I take `branch_name` as things like "master", or "origin/master". `branch` is still a reference under a "well-known" location. To me, `upstream_branch` refers to the end goal, and because it needs to be understood to be provided correctly (and given that branches are so transparent already), it's more confusing, as I think it's the first parameter that takes a "branch name" without actually being named `branch_name`. Thanks for the feedback though, I have some good documentation fodder now for naming things correctly 😉.